storage/innobase/row/row0merge.cc

```diff
@@ -2238,6 +2240,17 @@ row_merge_read_clustered_index(
 			ut_ad(add_autoinc
 			      < dict_table_get_n_user_cols(new_table));
 
+			bool historical_row = false;
+			if (new_table->versioned()) {
+				const dfield_t *dfield = dtuple_get_nth_field(
+					row, new_table->vers_end);
+				const byte *data = static_cast<const byte *>(
+					dfield_get_data(dfield));
+				ut_ad(dfield_get_len(dfield) == 8);
+				historical_row =
+					mach_read_from_8(data) != TRX_ID_MAX;
```

Comment:
Can we add a predicate `dfield_t::is_historical_row()` or similar? There is no need to decode `mach_read_from_8()` here. A simple `memcmp()` against eight bytes of `0xff` is faster.